api/kv_v2.go

```diff
@@ -399,3 +493,87 @@ func extractFullMetadata(secret *Secret) (*KVMetadata, error) {
 
 	return metadata, nil
 }
+
+func mergePatch(ctx context.Context, client *Client, path string, wrappedData map[string]interface{}) (*Secret, error) {
+	secret, err := client.Logical().JSONMergePatch(ctx, path, wrappedData)
+	if err != nil {
+		// If it's a 405, that probably means the server is running a pre-1.9
+		// Vault version that doesn't support the HTTP PATCH method.
+		// Fall back to the old way of doing it.
+		if re, ok := err.(*ResponseError); ok && re.StatusCode == 405 {
+			return readThenWrite(ctx, client, path, wrappedData)
+		}
+
+		if re, ok := err.(*ResponseError); ok && re.StatusCode == 403 {
+			return nil, fmt.Errorf("received 403 from Vault server; please ensure that token's policy has \"patch\" capability: %w", err)
```

Comment:
Whenever you get a chance to review this @ccapurso , I wanted to ask you if this seems like okay behavior for the `api` package version of this error-handling scenario. 

In the `command` package's version of this method [here](https://github.com/hashicorp/vault/blob/main/command/kv_patch.go#L345-L348), you can see that it doesn't error in this scenario, it throws a warning and resorts back to "rw" style updates. Since it's hard to throw a "warning" in the client library, I thought it might be good to simplify this logic and just error out like I have here. Note that I removed the `rwFallback` variable altogether. However, I can see an argument to be made that these helpers should work the same way as it does in the CLI, so... let me know what you think.

(In case you're wondering why I didn't just reuse those methods altogether, I opted to not share code between api/ and command/ because there is some terminal-printing-specific code in command/ that might be hard to untangle, and we plan on moving the api repo out of this monorepo in the near-ish future anyway.)

Reply:
Another idea @VinnyHC had was to perhaps simplify this altogether and make this Patch helper only capable of one style of Patch request. But that would mean either being exclusionary to pre-1.9.0 users or not utilizing the JSON merge patch style... so we would appreciate your context!

Reply:
Apologies, I wasn't very clear; my question was around the behavior/signature of the function `Patch` and if we should explicitly require a method to be passed in or have a default with a mechanism to override it.
The current code relies on a default value, where we allow the users to alter the default method by passing in the `WithMethod` KVOption. Should we be more forceful/explicit and have it as an argument in `Patch`?
Like something similar to this:
```go
func (kv *kvv2) Patch(ctx context.Context, secretPath string, data map[string]interface{}, mergeMethod string, opts ...KVOption) {...}
```
Further, if we do the above option would we want to implement something like:
```go
type MergeMethod string

const(
	Patch MergeMethod = "patch"
	RW MergeMethod = "rw"
)

func (kv *kvv2) Patch(ctx context.Context, secretPath string, data map[string]interface{}, m MergeMethod, opts ...KVOption) {...}
```

Reply:
Consistency between the CLI is an interesting topic. The reason the warning exists and a fallback occurs is that the `vault kv patch` command preexisted support for the HTTP `PATCH` method.  Vault needed to provide proper HTTP `PATCH` support via the CLI while ensuring backward compatibility to ensure we didn't break existing workflows. I am not sure we need to do the same here, however. There is less likelihood for someone to take proper action given the programmatic use. With that said, I think that erroring here makes sense.

Reply:
I see, thanks for the context. I'll keep the RW functionality available this time since I've already coded it here and there's no sense in deleting it if it's helpful for pre-1.9.0 users, but next time maybe I don't need to worry about old Vault servers from before two server versions ago when writing new functions like this.